grakn-graph/src/main/java/ai/grakn/graph/internal/concept/RelationReified.java

```diff
@@ -89,34 +92,60 @@ public void addRolePlayer(Role role, Thing thing) {
     /**
      * Sets the internal hash in order to perform a faster lookup
      */
-    public void setHash(){
-        vertex().propertyUnique(Schema.VertexProperty.INDEX, generateNewHash(type(), allRolePlayers()));
+    public void setHash(String hash){
+        vertex().propertyUnique(Schema.VertexProperty.INDEX, hash);
     }
 
     /**
      *
      * @param relationType The type of this relation
      * @param roleMap The roles and their corresponding role players
-     * @return A unique hash identifying this relation
+     * @return A unique hash identifying this {@link Relation}
      */
     public static String generateNewHash(RelationType relationType, Map<Role, Set<Thing>> roleMap){
         SortedSet<Role> sortedRoleIds = new TreeSet<>(roleMap.keySet());
         StringBuilder hash = new StringBuilder();
-        hash.append("RelationType_").append(relationType.getId().getValue().replace("_", "\\_")).append("_Relation");
+        hash.append("RelationType_").append(StringUtil.escapeString(relationType.getId().getValue())).append("_Relation");
 
         for(Role role: sortedRoleIds){
-            hash.append("_").append(role.getId().getValue().replace("_", "\\_"));
+            hash.append("_").append(StringUtil.escapeString(role.getId().getValue()));
 
             roleMap.get(role).forEach(instance -> {
                 if(instance != null){
-                    hash.append("_").append(instance.getId().getValue().replace("_", "\\_"));
+                    hash.append("_").append(StringUtil.escapeString(instance.getId().getValue()));
                 }
             });
         }
         return hash.toString();
     }
 
     /**
+     * Creates a hash for a relation based on it's {@link RelationType} and the {@link Resource} which serves as it's key
+     *
+     * @param relationType the {@link RelationType} of the {@link Relation}
+     * @param resourceMap a sorted map of {@link ai.grakn.concept.ResourceType} Ids to {@link Resource} Ids
+     * @return A unique hash identifying this {@link Relation}
+     */
+    public static String generateNewHash(RelationType relationType, TreeMap<String, String> resourceMap){
```

Comment:
Good questions. It's actually an id composed of several other ids. I think I landed on using the word "hash" because "id" and "Index" were already taken in different contexts. Maybe "InternalID" would be more clear?